### ISSUE TITLE
feature request: allow adding an 'Explicit' attribute to WAT tests

### ISSUE THREAD

**Issue by pracplayopen** [AlexDarigan/WAT#312]
DESCRIPTION OF REQUEST

Presently tests are identified with the [Test] attribute.

Would like an additional optional attribute to be supported to allow explicit, eg:

possibility 1:  [Test,Explicit]
possibility 2: [Test(Explicit=true)]

PURPOSE

To allow tests to be created and run, but flagged as not being part of 'All' group.

This allows clean CD/CI pipeline usage, as tests in development can be tagged as explicit and run manually until read for integration/production.

Also allows seperation between always-run tests and useful-but-not-always-run tests (these can be tests that are extremely costly in bandwidth, $, cpu, time or other resources such that they can't be run every time, but are still useful to have eeasily accessible.

DISCUSSION

While in theory you could use WAT tags for this (since the cli supports running only certain tags), tags are less than ideal for several reasons:

1. tags are only editable in the gui, not in code
2. there is no default 'not-all' tag, such as explicit would allow... to simpliy exclude from the automatic 'all' group
3. tagging isn't done 'near' the test, which makes it potentially more costly to maintain as people changing code have c# skill and maynot be godot experts

You might also solve this by moving tagging to the test attribute, tho haviing a single explicit attribute seems less work to implement and also.... more explicit about it's explicitly. :)

**Comment by RailKill** [comment]
I second this suggestion. If I may, I would like to further improve this specification as I think there are a few other cases that should be covered in terms of specifying when a test should or should not be covered in run=all. Perhaps something like "Mode" would be a more fitting name, I'm sure someone can suggest a better name, but this is what I can think of for now. By default, tests are picked up by run=all in all modes, but can be altered by the user with a 2 bit value:

C#
---
- ```[Test(Mode=0)]```: This test is excluded from run all. This is what we want, but we can have a few more options.
- ```[Test(Mode=1)]```: This test will only run in normal run all.
- ```[Test(Mode=2)]```: This test will only run in debug run all.
- ```[Test(Mode=3)]```: This test is included in any run all. This is the default value.

**Note:** This "Mode" check is for when ALL tests are being run. In any other case, it should be ignored.

If using integer values is not user-friendly, we can define these numbers as enums, so it will be something like (Mode=Exclude) which is equivalent to (Mode=0). Also, the [Test] attribute currently uses positional arguments for **parameterized tests**, but "Mode" is a named argument, so it can come after, like so:

`[Test("hello", 1, true, Mode=Exclude)]`

This also means that even certain parameters in a parameterized test can be excluded or included based on the Mode argument, very handy!


GDScript
---
I guess it'll have to be a method call inside the test method like ```set_mode(RunAll.EXCLUDE)``` similar to ```describe()``` or ```parameters()```.

**Comment by GlitchedCode** [comment]
> ## C#
> * `[Test(Mode=0)]`: This test is excluded from run all. This is what we want, but we can have a few more options.
> * `[Test(Mode=1)]`: This test will only run in normal run all.
> * `[Test(Mode=2)]`: This test will only run in debug run all.
> * `[Test(Mode=3)]`: This test is included in any run all. This is the default value.

No reason as to why the mode parameter shouldn't be an enum type in C# as well.

**Comment by AlexDarigan** [comment]
I'll keep this feature in mind for future but I'm effectivelly having WAT enter maintenance only at the moment, so all new features will be benched until then. For CI/CD usage, there are ways around this (some mentioned in the OP) so I hope you can be patient until then.

**Comment by AlexDarigan** [comment]
Reopening this as the request is unresolved.

**Comment by AlexDarigan** [comment]
> I second this suggestion. If I may, I would like to further improve this specification as I think there are a few other cases that should be covered in terms of specifying when a test should or should not be covered in run=all. Perhaps something like "Mode" would be a more fitting name, I'm sure someone can suggest a better name, but this is what I can think of for now. By default, tests are picked up by run=all in all modes, but can be altered by the user with a 2 bit value:
> 
> ## C#
> * `[Test(Mode=0)]`: This test is excluded from run all. This is what we want, but we can have a few more options.
> * `[Test(Mode=1)]`: This test will only run in normal run all.
> * `[Test(Mode=2)]`: This test will only run in debug run all.
> * `[Test(Mode=3)]`: This test is included in any run all. This is the default value.
> 
> **Note:** This "Mode" check is for when ALL tests are being run. In any other case, it should be ignored.
> 
> If using integer values is not user-friendly, we can define these numbers as enums, so it will be something like (Mode=Exclude) which is equivalent to (Mode=0). Also, the [Test] attribute currently uses positional arguments for **parameterized tests**, but "Mode" is a named argument, so it can come after, like so:
> 
> `[Test("hello", 1, true, Mode=Exclude)]`
> 
> This also means that even certain parameters in a parameterized test can be excluded or included based on the Mode argument, very handy!
> 
> ## GDScript
> I guess it'll have to be a method call inside the test method like `set_mode(RunAll.EXCLUDE)` similar to `describe()` or `parameters()`.

If people are still interested, I could add a run flag that you can check manually at the start of each test function.

```
[Test]
public void MyTestFunc() {
    if RUN_MODE == RUN_MODE.DEBUG:
        return
 ```
 
 Not ideal but a start until GDScript 4.0 gets custom annotations.

**Comment by AlexDarigan** [comment]
Presumably, tests should run despite this attribute if using any option that isn't run all?

**Comment by AlexDarigan** [comment]
Resolved by a0054a3dc259b0e74dbb6ba1bef6282bbf465eb6 - Added flag methods (best I can do in 3.X GDScript) for both GDScript and CSharp

**GDScript**
is_any_run_all()
is_normal_run_all()
is_debug_run_all()

**CSharp**
IsAnyRunAll()
IsNormalRunAll()
IsDebugRunAll()

Can't wait till 4.0 gets custom annotations.

**Comment by AlexDarigan** [comment]
Addendum: This does allow you to exit the test early but you may need to use auto-pass in contexts where it doesn't run.
